spring-integration-file/src/main/java/org/springframework/integration/file/filters/CompositeFileListFilter.java

```diff
@@ -101,6 +107,17 @@ public void close() throws IOException {
 		return this;
 	}
 
+	@Override
+	public void addDiscardCallback(DiscardCallback<F> discardCallback) {
+		this.discardCallback = discardCallback;
+		if (this.discardCallback != null) {
+			this.fileFilters
+					.stream()
+					.filter(DiscardAwareFileListFilter.class::isInstance)
+					.map(f -> (DiscardAwareFileListFilter<F>) f)
+					.forEach(f -> f.addDiscardCallback(discardCallback));
```

Comment:
We must only call the callback once per file - remember the composite filter passes all files to all filters.

Or, change the queue to linked set.

Reply:
Well, as far as `CompositeFileListFilter` calls `filterFiles()` for each its delegate, it really a part of the API that `discardCallback` will be called from them. But pay attention that this `CompositeFileListFilter` doesn't call `discardCallback` - it just populates to the delegates.

But I agree: let's reconsider the `WatchServiceScanner` to use a some collection with uniqueness to avoid a double call for the same file from different filters!